Jenkinsfile

```diff
@@ -63,6 +63,7 @@ node {
                 }
             } finally {
                 postgresContainer.stop()
+                cleanWs()
```

Comment:
All the testing-related build stages in this file run in parallel - see the `parallel failFast: true` line above. Does each stage use a separate workspace, or do they share one, in which case removing the workspace at the end of any one of the stages could affect the others.

Reply:
I see, that's not an issue we've had to think about for h yet. The cleanWs() should be moved so that it happens after the last stage. I'll try and find an answer in the docs, but if documentation is lacking it may come down to "try it and see".

Reply:
It looks to me like the way this has been done in h's `Jenkinsfile` isn't quite right. `cleanWs()` has been added to the `finally` of a `try` that contains only _part_ of the pipeline. If the Jenkins job fails before it gets into this `try` then `cleanWs()` won't be run.

In both h and lms I think you'll need to wrap the entire contents of the `node { ...}` in a `try ... finally`. See the cleanup and notifications docs here (and click the _Toggle Scripted Pipeline_ button): https://jenkins.io/doc/pipeline/tour/post/. So something like:

```groovy
node {
  try { 
    runArgs = "-u root -e SITE_PACKAGES=true"

    stage("Build") {
       ...
    }

    // Run each of the stages in parallel.
    parallel failFast: true,
    "Formatting": {
      ...
        }
    },
    ...
    "Frontend tests": {
      ...
    }

    onlyOnMaster {
      ...
    }

  } finally {
    cleanWs()
  }
}

onlyOnMaster {
    milestone()
    stage("qa deploy") {
      ...
}
...
```

Reply:
The deploy to QA and production are outside of the `node { ...}`, and therefore outside of the `try { ... } finally { ... }`, so I think the `cleanWs()` is getting executed **before** the deployment to QA and prod happens. This is currently the case on h master. This _seems_ to be okay -- I guess the deploy to QA and prod doesn't need anything in the workspace that gets cleaned up (and also doesn't create any new stuff in the workspace, post-clean, that then _isn't_ getting cleaned?)